### PR TITLE
Create sample spring-cloud-gateway + AAD resource server

### PR DESCRIPTION
**Pull request by chenrujun**
As title.

### CODE REVIEW

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/client/configuration/ApplicationConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.security.oauth2.client.registration.ReactiveClientRegistrationRepository;
+import org.springframework.security.oauth2.client.web.reactive.function.client.ServerOAuth2AuthorizedClientExchangeFilterFunction;
+import org.springframework.security.oauth2.client.web.server.ServerOAuth2AuthorizedClientRepository;
+import org.springframework.web.reactive.function.client.WebClient;
+
+@Configuration
+public class ApplicationConfiguration {
+
+    @Bean
+    public static WebClient webClient(ReactiveClientRegistrationRepository clientRegistrationRepository,
```
Comment (backwind1233) [inline aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/client/configuration/ApplicationConfiguration.java:14]: why is static here?

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/client/configuration/SecurityConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+@EnableWebFluxSecurity
+public class SecurityConfiguration {
+
+    @Bean
+    SecurityWebFilterChain springSecurityFilterChain(ServerHttpSecurity http) {
+        http.oauth2Client();
+        return http.build();
+    }
+
+}
```
Comment (backwind1233) [inline aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/client/configuration/SecurityConfiguration.java:17]: add an empty line?

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/client/controller/ResourceServer1Controller.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+public class ResourceServer1Controller {
+
+    private final WebClient webClient;
+
+    public ResourceServer1Controller(WebClient webClient) {
+        this.webClient = webClient;
+    }
+
+    @GetMapping("/resource-server-1")
+    public Mono<String> hello(@RegisteredOAuth2AuthorizedClient("client-1-resource-server-1") OAuth2AuthorizedClient client1ResourceServer1) {
```
Comment (backwind1233) [inline aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/client/controller/ResourceServer1Controller.java:22]: I think it's too long here for code, we don't have checkstyle now, but I think we will.
The code line better not longer than 100.

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/client/controller/ResourceServer2Controller.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+public class ResourceServer2Controller {
+
+    private final WebClient webClient;
+
+    public ResourceServer2Controller(WebClient webClient) {
+        this.webClient = webClient;
+    }
+
+    @GetMapping("/resource-server-2")
+    public Mono<String> hello(@RegisteredOAuth2AuthorizedClient("client-1-resource-server-2") OAuth2AuthorizedClient client1ResourceServer1) {
```
Comment (backwind1233) [inline aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/client/controller/ResourceServer2Controller.java:22]: same here

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/resources/application.yml [hunk 5]
```diff
@@ -0,0 +1,35 @@
+# Please fill these placeholders before run this application:
```
Comment (backwind1233) [inline aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/resources/application.yml:1]: run -> running

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/gateway/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/gateway/configuration/ApplicationConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Predicate;
+
+@Configuration
+public class ApplicationConfiguration {
+
+    @Value("${spring.security.oauth2.resourceserver.jwt.audiences}")
+    List<String> audiences;
```
Comment (backwind1233) [inline aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/gateway/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/gateway/configuration/ApplicationConfiguration.java:27]: add private

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/gateway/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/gateway/configuration/SecurityConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package com.azure.spring.sample.activedirectory.reactive.webflux.oauth2.gateway.gateway.configuration;
+
+import org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurity;
+
+@EnableWebFluxSecurity
+public class SecurityConfiguration {
```
Comment (backwind1233) [inline aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/gateway/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/gateway/configuration/SecurityConfiguration.java:6]: what's the difference with ApplicationConfiguration, can we put the @EnableWebFluxSecurity in the same class?

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/resource-server-1/src/main/java/com/azure/spring/sample/activedirectory/oauth2/gateway/resourceserver1/configuration/ApplicationConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Predicate;
+
+@Configuration
+public class ApplicationConfiguration {
+
+    @Value("${spring.security.oauth2.resourceserver.jwt.audiences}")
+    List<String> audiences;
```
Comment (backwind1233) [inline aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/resource-server-1/src/main/java/com/azure/spring/sample/activedirectory/oauth2/gateway/resourceserver1/configuration/ApplicationConfiguration.java:27]: private

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/resource-server-2/src/main/resources/application.yml [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+server:
+  port: 8082
+spring:
+  security:
+    oauth2:
+      resourceserver:
+        jwt:
+          jwk-set-uri: https://login.microsoftonline.com/${tenant-id}/discovery/v2.0/keys
+          issuer-uri: https://login.microsoftonline.com/${tenant-id}/v2.0
+          audiences: ${resource-server-2-client-id}
```
Comment (backwind1233) [inline aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/resource-server-2/src/main/resources/application.yml:17]: I suggest using the upper case here for variables in yaml, 
${resource-server-2-client-id} --> ${RESOURCE-SERVER-2-CLIENT-ID}

1. our spring cloud [azure](https://microsoft.github.io/spring-cloud-azure/docs/current/reference/html/index.html#auto-configure-azure-sdk-clients) doc follow this pattern.
2.other samples I refactored using the upper case.
it's better to be consistent.

File: aad/spring-security/servlet/oauth2/client-access-multiple-resource-server/client/src/main/resources/application.yml [hunk 10]
```diff
@@ -32,4 +32,6 @@ spring:
             client-id: ${client-1-client-id}
             client-secret: ${client-1-client-secret}
             scope: openid, profile, api://${resource-server-2-client-id}/resource-server-2.scope-1 # Refs: https://docs.microsoft.com/azure/active-directory/develop/v2-permissions-and-consent#openid-connect-scopes
-            redirect-uri: http://localhost:8080/login/oauth2/code/
\ No newline at end of file
+            redirect-uri: http://localhost:8080/login/oauth2/code/
+  profiles:
+    active: develop
```
Comment (backwind1233) [inline aad/spring-security/servlet/oauth2/client-access-multiple-resource-server/client/src/main/resources/application.yml:37]: Add an empty line?

File: aad/spring-security/servlet/oauth2/client-access-multiple-resource-server/resource-server-1/pom.xml [hunk 11]
```diff
@@ -11,7 +11,7 @@
     <relativePath>../../../../../../pom.xml</relativePath>
   </parent>
 
-  <artifactId>servlet-oauth2-client-access-multiple-resource-server-resourc-server-1-application</artifactId>
+  <artifactId>servlet-oauth2-client-access-multiple-resource-server-resource-server-1-application</artifactId>
```
Comment (backwind1233) [inline aad/spring-security/servlet/oauth2/client-access-multiple-resource-server/resource-server-1/pom.xml:14]: If you change this(from `resourc-` to `resource-`) here in this PR, do you think we should change all of them?
image

**Review by backwind1233 — COMMENTED** [review]
🐮

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/client/configuration/ApplicationConfiguration.java (hunk 1 above)
Comment (chenrujun) [reply aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/client/configuration/ApplicationConfiguration.java]: Updated.

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/client/configuration/SecurityConfiguration.java (hunk 2 above)
Comment (chenrujun) [reply aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/client/configuration/SecurityConfiguration.java]: Updated.

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/client/controller/ResourceServer1Controller.java (hunk 3 above)
Comment (chenrujun) [reply aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/client/controller/ResourceServer1Controller.java]: Updated.

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/client/controller/ResourceServer2Controller.java (hunk 4 above)
Comment (chenrujun) [reply aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/client/controller/ResourceServer2Controller.java]: Updated.

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/resources/application.yml (hunk 5 above)
Comment (chenrujun) [reply aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/client/src/main/resources/application.yml]: Updated.

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/gateway/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/gateway/configuration/ApplicationConfiguration.java (hunk 6 above)
Comment (chenrujun) [reply aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/gateway/src/main/java/com/azure/spring/sample/activedirectory/reactive/webflux/oauth2/gateway/gateway/configuration/ApplicationConfiguration.java]: Updated.

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/resource-server-1/src/main/java/com/azure/spring/sample/activedirectory/oauth2/gateway/resourceserver1/configuration/ApplicationConfiguration.java (hunk 8 above)
Comment (chenrujun) [reply aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/resource-server-1/src/main/java/com/azure/spring/sample/activedirectory/oauth2/gateway/resourceserver1/configuration/ApplicationConfiguration.java]: Updated.

File: aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/resource-server-2/src/main/resources/application.yml (hunk 9 above)
Comment (chenrujun) [reply aad/spring-security/reactive/webflux/oauth2/spring-cloud-gateway/resource-server-2/src/main/resources/application.yml]: Updated.

File: aad/spring-security/servlet/oauth2/client-access-multiple-resource-server/client/src/main/resources/application.yml (hunk 10 above)
Comment (chenrujun) [reply aad/spring-security/servlet/oauth2/client-access-multiple-resource-server/client/src/main/resources/application.yml]: Updated.

File: aad/spring-security/servlet/oauth2/client-access-multiple-resource-server/resource-server-1/pom.xml (hunk 11 above)
Comment (chenrujun) [reply aad/spring-security/servlet/oauth2/client-access-multiple-resource-server/resource-server-1/pom.xml:14]: Updated.

**Review by backwind1233 — APPROVED** [review]
LGTM

**Pull request merged**